File: src/lib/providers/anime-episode-enrich.ts
import { kitsuToMal, kitsuToTvdb } from "@/lib/providers/anime-mapping";
import { harborImdbEpisodes } from "@/lib/providers/harbor-imdb";
import { fillerEpisodes } from "@/lib/anime-fillers";
import { fetchTvdbThumbs } from "@/lib/providers/anime-tvdb-thumbs";
import { meta as fetchCinemetaMeta } from "@/lib/cinemeta";
import type { KitsuEpisode } from "@/lib/providers/kitsu";
import type { Settings } from "@/lib/settings";
import { mergeCinemetaEpisodes } from "./anime-episode-cinemeta-merge";

async function enrichFiller(episodes: KitsuEpisode[], kitsuId: number): Promise<void> {
  if (episodes.some((ep) => ep.filler)) return;
  const malId = await kitsuToMal(kitsuId).catch(() => null);
  if (!malId) return;
  const fillers = await fillerEpisodes(malId).catch(() => new Set<number>());
  if (fillers.size === 0) return;
  for (const ep of episodes) {
    const num = ep.absoluteNumber ?? ep.number;
    if (fillers.has(num)) ep.filler = true;
  }
}

async function enrichCinemetaEpisodes(
  episodes: KitsuEpisode[],
  imdbId: string | null,
): Promise<void> {
  if (!imdbId || !imdbId.startsWith("tt")) return;
  if (
    episodes.every(
      (ep) => ep.thumbnail && ep.synopsis && ep.title && ep.title !== `Episode ${ep.number}`,
    )
  )
    return;
  const m = await fetchCinemetaMeta("series", imdbId).catch(() => null);
  const videos = m?.videos ?? [];
  if (videos.length === 0) return;
  mergeCinemetaEpisodes(episodes, videos);
}

async function enrichTvdbThumbs(
  episodes: KitsuEpisode[],
  settings: Settings,
  kitsuId: number,
): Promise<void> {
  if (!settings.tvdbKey) return;
  if (episodes.every((ep) => ep.thumbnail)) return;
  const tvdbId = await kitsuToTvdb(kitsuId).catch(() => null);
  if (!tvdbId) return;
  const seasons = Array.from(new Set(episodes.map((ep) => ep.imdbSeason ?? ep.seasonNumber ?? 1)));
  const index = await fetchTvdbThumbs(settings.tvdbKey, tvdbId, seasons).catch(() => null);
  if (!index) return;
  for (const ep of episodes) {
    if (ep.thumbnail) continue;
    const season = ep.imdbSeason ?? ep.seasonNumber ?? 1;
    const epNum = ep.imdbEpisode ?? ep.number;
    const hit =
      index.bySeasonEpisode.get(`${season}:${epNum}`) ??
      (ep.absoluteNumber ? index.byAbsolute.get(ep.absoluteNumber) : undefined) ??
      index.byAbsolute.get(ep.number);
    if (hit) ep.thumbnail = hit;
  }
}

async function enrichHarborImdb(episodes: KitsuEpisode[], imdbId: string | null): Promise<void> {
  if (!imdbId || !imdbId.startsWith("tt")) return;
  const map = await harborImdbEpisodes(imdbId).catch(() => null);
  if (!map || map.size === 0) return;
  for (const ep of episodes) {
    const season = ep.imdbSeason ?? ep.seasonNumber ?? 1;
    const num = ep.imdbEpisode ?? ep.number;
    const real = map.get(`${season}:${num}`);
    if (real != null && real > 0) {
      ep.rating = real;
      ep.ratingIsImdb = true;
    }
  }
}

export async function enrichEpisodes(
  episodes: KitsuEpisode[],
  settings: Settings,
  kitsuId: number,
  imdbId: string | null = null,
): Promise<void> {
  await Promise.all([
    enrichFiller(episodes, kitsuId),
    enrichHarborImdb(episodes, imdbId),
    (async () => {
      await enrichCinemetaEpisodes(episodes, imdbId);
      await enrichTvdbThumbs(episodes, settings, kitsuId);
    })(),
  ]);
}

File: clients/lib/domain/anime/anime_episode_enrich.dart
import '../../core/http/json_transport.dart';
import '../addons/addon_client.dart';
import '../addons/models.dart';
import '../catalog/cinemeta.dart';
import '../catalog/tvdb.dart';
import '../ratings/harbor_imdb.dart';
import 'anime_fillers.dart';
import 'anime_mapping.dart';
import 'anime_tvdb_thumbs.dart';
import 'kitsu_client.dart';

bool _hasThumb(KitsuEpisode ep) => ep.thumbnail?.isNotEmpty ?? false;

/// Enriches an anime's episodes in place with data the base Kitsu/addon build
/// lacks: filler flags, fresh IMDb episode ratings, and episode thumbnails from
/// Cinemeta then TVDB. Ported 1:1 from `anime-episode-enrich.ts`. Each source is
/// best-effort — a failure leaves the episodes as they were.
class AnimeEpisodeEnricher {
  AnimeEpisodeEnricher({
    required AnimeMapper mapper,
    required AnimeFillers fillers,
    required AddonClient addon,
    required TvdbClient tvdb,
    required JsonTransport transport,
  }) : _mapper = mapper,
       _fillers = fillers,
       _addon = addon,
       _tvdb = tvdb,
       _transport = transport;

  final AnimeMapper _mapper;
  final AnimeFillers _fillers;
  final AddonClient _addon;
  final TvdbClient _tvdb;
  final JsonTransport _transport;

  /// Runs the filler, IMDb-rating and thumbnail enrichers over [episodes]. The
  /// filler and rating passes run concurrently with the thumbnail pass, which
  /// tries Cinemeta before falling back to TVDB.
  Future<void> enrich(
    List<KitsuEpisode> episodes, {
    required int kitsuId,
    String? imdbId,
    required String tvdbKey,
  }) async {
    await Future.wait([
      _enrichFiller(episodes, kitsuId),
      _enrichHarborImdb(episodes, imdbId),
      _enrichThumbs(episodes, imdbId, kitsuId, tvdbKey),
    ]);
  }

  Future<void> _enrichFiller(List<KitsuEpisode> episodes, int kitsuId) async {
    if (episodes.any((ep) => ep.filler == true)) return;
    final malId = await _guard(_mapper.kitsuToMal(kitsuId));
    if (malId == null || malId == 0) return;
    final fillers =
        await _guard(_fillers.fillerEpisodes(malId)) ?? const <int>{};
    if (fillers.isEmpty) return;
    for (final ep in episodes) {
      final n = ep.absoluteNumber ?? ep.number;
      if (fillers.contains(n)) ep.filler = true;
    }
  }

  Future<void> _enrichHarborImdb(
    List<KitsuEpisode> episodes,
    String? imdbId,
  ) async {
    if (imdbId == null || !imdbId.startsWith('tt')) return;
    final map = await _guard(harborImdbEpisodes(_transport, imdbId));
    if (map == null || map.isEmpty) return;
    for (final ep in episodes) {
      final season = ep.imdbSeason ?? ep.seasonNumber;
      final n = ep.imdbEpisode ?? ep.number;
      final real = map['$season:$n'];
      if (real != null && real > 0) {
        ep.rating = real;
        ep.ratingIsImdb = true;
      }
    }
  }

  Future<void> _enrichThumbs(
    List<KitsuEpisode> episodes,
    String? imdbId,
    int kitsuId,
    String tvdbKey,
  ) async {
    await _enrichCinemetaThumbs(episodes, imdbId);
    await _enrichTvdbThumbs(episodes, kitsuId, tvdbKey);
  }

  Future<void> _enrichCinemetaThumbs(
    List<KitsuEpisode> episodes,
    String? imdbId,
  ) async {
    if (imdbId == null || !imdbId.startsWith('tt')) return;
    if (episodes.every(_hasThumb)) return;
    final meta = await _guard(_cinemetaSeries(imdbId));
    final videos = meta?.videos ?? const <VideoRef>[];
    if (videos.isEmpty) return;

    final bySeasonEpisode = <String, String>{};
    final byAbsolute = <int, String>{};
    final ordered =
        [
          for (final v in videos)
            if ((v.thumbnail?.isNotEmpty ?? false) &&
                v.season != null &&
                v.episode != null)
              v,
        ]..sort((a, b) {
          final s = (a.season ?? 0).compareTo(b.season ?? 0);
          return s != 0 ? s : (a.episode ?? 0).compareTo(b.episode ?? 0);
        });
    var pos = 0;
    for (final v in ordered) {
      final thumb = v.thumbnail!;
      bySeasonEpisode['${v.season}:${v.episode}'] = thumb;
      if ((v.season ?? 0) > 0) {
        pos += 1;
        byAbsolute.putIfAbsent(pos, () => thumb);
      }
    }

    for (final ep in episodes) {
      if (_hasThumb(ep)) continue;
      final season = ep.imdbSeason ?? ep.seasonNumber;
      final epNum = ep.imdbEpisode ?? ep.number;
      final hit =
          bySeasonEpisode['$season:$epNum'] ??
          byAbsolute[ep.absoluteNumber ?? ep.number];
      if (hit != null) ep.thumbnail = hit;
    }
  }

  Future<void> _enrichTvdbThumbs(
    List<KitsuEpisode> episodes,
    int kitsuId,
    String tvdbKey,
  ) async {
    if (tvdbKey.isEmpty) return;
    if (episodes.every(_hasThumb)) return;
    final tvdbId = await _guard(_mapper.kitsuToTvdb(kitsuId));
    if (tvdbId == null || tvdbId == 0) return;
    final seasons = <int>{
      for (final ep in episodes) ep.imdbSeason ?? ep.seasonNumber,
    }.toList();
    final index = await _guard(
      fetchTvdbThumbs(_tvdb, tvdbKey, tvdbId, seasons),
    );
    if (index == null) return;
    for (final ep in episodes) {
      if (_hasThumb(ep)) continue;
      final season = ep.imdbSeason ?? ep.seasonNumber;
      final epNum = ep.imdbEpisode ?? ep.number;
      final abs = ep.absoluteNumber;
      final hit =
          index.bySeasonEpisode['$season:$epNum'] ??
          (abs != null ? index.byAbsolute[abs] : null) ??
          index.byAbsolute[ep.number];
      if (hit != null) ep.thumbnail = hit;
    }
  }

  Future<Meta?> _cinemetaSeries(String imdbId) async =>
      (await _addon.meta(cinemetaBase, 'series', imdbId)).valueOrNull;

  static Future<T?> _guard<T>(Future<T> f) async {
    try {
      return await f;
    } catch (_) {
      return null;
    }
  }
}
